blog/posts/announcing-melange-5.md

```diff
@@ -0,0 +1,285 @@
+---
+title: Announcing Melange 5
+date: 2025-03-05
+author: Antonio Monteiro
+gravatar: 45c2052f50f561b9dc2cae59c777aecd794f57269fa317f9c9c3365c2e00d16f
+twitter: '@_anmonteiro'
+---
+
+We are excited to announce the release of Melange 5, the compiler for OCaml
+that targets JavaScript.
+
+A lot of goodies went into this release! While our focus was mostly on features
+that make it easy to express more JavaScript constructs and supporting OCaml
+5.3, we also managed to fit additional improvements in the release: better
+editor support for Melange `external`s, code generation improvements, and
+better compiler output for generated JS. The most notable feature we're
+shipping in Melange 5 is support for JavaScript's [dynamic
+`import()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import),
+which we'll describe in detail below.
+
+Read on for the highlights.
+
+## Dynamic `import()` without sacrificing type safety
+
+Support for JavaScript's dynamic `import()` is probably what I'm most excited
+about in this Melange release. In Melange 5, we're releasing support for
+JavaScript's `import()` via a new function in `melange.js`, `Js.import: 'a ->
+'a promise`. I gave a small preview of dynamic `import()` during my [Melange
+talk](https://www.youtube.com/watch?v=3oCXT-ycHHs) at [Fun OCaml
+2024](https://fun-ocaml.com).
+
+`Js.import` is **type-safe** and **build system-compatible**. Let's break it
+down:
+
+1. **build system-compatible**: dynamic `import()`s in Melange work with Dune
+   out of the box: as usual, you must specify your `(library ..)` dependencies
+   in the `dune` file. At compile time, Melange will be aware of the
+   dynamically imported module locations to emit the arguments to
+   `import("/path/to/module.js)` automatically.
+2. **type-safe**: your OCaml code is still aware of its dependencies at
+   compile-time, but Melange will skip emitting static JavaScript `import ..`
+   declarations if the values are exclusively used through `Js.import(..)`.
+
+### Dynamically `import()`ing OCaml code
+
+The example below makes it clear: we import the entire `Stdlib.Int` module,
+specify its type signature, and observe that no static `import`s appear in the
```

Comment:
Should we clarify (here or somewhere else) that dynamic imports are an "unsafe" boundary, similar to bindings? Meaning, the developer is responsible of defining the right module type being imported, and the compiler has no way of checking if this definition is correct.

Reply:
that's actually wrong :) it's fully type-safe, and the compiler would complain if `module Int` wasn't of `module type of int`!